### ISSUE TITLE
PgSync Daemon is not clearing the cache items

### ISSUE THREAD

**Issue by CE-Creates** [toluaina/pgsync#86]
PGSync version: 1.1.24

Postgres version: 11.8

Elasticsearch version: 
v7.10.0

Python version: 3.7.9

Problem Description: 

The replication slot keep growing at times and when we look at the pgsync daemon logs, we could see lots of cache items pending to process. 

This problem is happening only when we have some bulk update done to the database by a script in the backend. 
This script update almost 80% of the data which could be what is creating this cache items. 

However, it never clear it out. and we always had to delete the replication slot and bootstrap again to restore.
This is also causing spike in RDS at times. 

Error Message (if any): 

```
Polling db my_db: 40,385 cache items
```

**Comment by andresrojasbimbau** [comment]
happening to me as well! thanks for reporting @CE-Creates, RDS has to scale up due to the amount of storage required

**Comment by toluaina** [comment]
1. Did you setup `REPLICATION_SLOT_CLEANUP_INTERVAL` as described [here](https://pgsync.com/env-vars/) ?
2. If you did the above do you have either `POLL_TIMEOUT` or `REDIS_POLL_INTERVAL` set to a sufficiently high value.
3. Maybe you can turn on logging, (or run with `-v`) set `REPLICATION_SLOT_CLEANUP_INTERVAL` to a low value and check if the cleanup is being triggered at all

**Comment by toluaina** [comment]
PS: Cache items is a bit misleading. This is not items to process but items that **_HAVE_** been processed

**Comment by toluaina** [comment]
are you running pgsync as a cron or a daemon? if its the latter you might want to schedule pgsync to run closer to the time you perform your bulk data load

**Comment by CE-Creates** [comment]
> are you running pgsync as a cron or a daemon? if its the latter you might want to schedule pgsync to run closer to the time you perform your bulk data load

We are running it as a daemon

**Comment by CE-Creates** [comment]
> PS: Cache items is a bit misleading. This is not items to process but items that **_HAVE_** been processed

Is there a metric that shows how many items are pending to be synced.
Till now, we were referring to this cache items value for that. 

During this period, the db transaction log spiked from 2.3 GB to 9.1 GB.
If cache items count is that of "have been processed", why is replication slot size not reducing?


Polling db my_db: 40,385 cache items
Polling db my_db: 40,385 cache items

**Comment by CE-Creates** [comment]
> 1. Did you setup `REPLICATION_SLOT_CLEANUP_INTERVAL` as described [here](https://pgsync.com/env-vars/) ?
> 2. If you did the above do you have either `POLL_TIMEOUT` or `REDIS_POLL_INTERVAL` set to a sufficiently high value.
> 3. Maybe you can turn on logging, (or run with `-v`) set `REPLICATION_SLOT_CLEANUP_INTERVAL` to a low value and check if the cleanup is being triggered at all

We have the following configured

REPLICATION_SLOT_CLEANUP_INTERVAL = 900
POLL_TIMEOUT = 0.5

We have not set REDIS_POLL_INTERVAL so I assume it must be using a default value

**Comment by toluaina** [comment]
- I see you are using a `REPLICATION_SLOT_CLEANUP_INTERVAL` of 15 mins(_default_).
- Try dropping that down a to a lower interval. 
- I will change the default to 5 mins.
- In any case bursts of bulk data imports will result in this 
- clearing the replication slot manually is not ideal. Its better to set it to a low interval

**Comment by CE-Creates** [comment]
Thank you @toluaina 
I did the following today:

1. set REPLICATION_SLOT_CLEANUP_INTERVAL to 300 ( 5 minutes )
2. deleted replication slot as it was getting huge
3. bootstrap
4. ran the daemon

I could see the following now in the log ( 0 cache items )

```
Polling db my_db: 0 cache items
Polling db my_db: 0 cache items
Polling db my_db: 0 cache items
Polling db my_db: 0 cache items
Polling db my_db: 0 cache items 
```

So far so good. But I am sure it will burst every time the script i mentioned above run .
Ideally we used to run it everyday morning, now we are running it manually to have a control of this issue.

**Comment by CE-Creates** [comment]
@toluaina 
I just noticed something today morning
I ran a script this morning which changed a lot of data (30% of the database entries).
As expected, pgsync log was showing cache items
Polling db my_db: 23.162 cache items

I was expecting it to be down to 0 but it didnt happen. 
the replication slot went up to 2 GB

I decided to stop pgsync and start the daemon again.
When restarted, the replication slot got cleared up and the log started showing 0 cache items

**Comment by toluaina** [comment]
As mentioned before, cache items are the items processed so far so there is nothing abnormal there.

The replication slot gets cleanup up at intervasl defined by `REPLICATION_SLOT_CLEANUP_INTERVAL`

**Comment by toluaina** [comment]
Closing this too. Feel free to re-open if this issue is still an issue :)

**Comment by smokedsalmonbagel** [comment]
I am also experiencing this.  I found a few related issues but this one is probably most similar.  I can sync if I periodically run pgsync but in daemon mode the 'cache items' increase but no changes are reflected in ES.  In my case I noticed one document with the same id showing up  once a second:
```
2021-05-06 23:07:57.807:DEBUG:pgsync.sync: on_notify: {'xmin': 60518827, 'new': {'id': 'xxxxxx-cd3b-41d2-a8ea-aef7aa798246', 'operator_id': None}, 'old': {'id': 'xxxxxxx-cd3b-41d2-a8ea-aef7aa79
8246', 'operator_id': None}, 'tg_op': 'UPDATE', 'table': 'users', 'schema': 'public'}
```
Its always the same doc _id.  Not sure if this is related or just some continuous DB activity.

I have tried changing some of the ENV VARs like REPLICATION_SLOT_CLEANUP_INTERVAL but I can't seem to get changes to show up in ES until I run it manually.
